### ISSUE TITLE
Separation of buffer data and client data

### ISSUE THREAD

**Issue by AljoschaMeyer** [google/xi-editor#125]
Let me preface this by saying that I don't have a background in the stuff I wrote about here, its just an intuitive brain dump. So I apologize in advance if this issue is

- wrong
- irrelevant
- out of scope
- uses way too many words
- and/or uses unclear and non-standard terminology

- - -

Right now, the RPC protocol from front-end to back-end tightly couples one client to one server. Just as there is exactly one text being operated upon (per tab), there is also exactly one viewport, one cursor etc. Is this considered a fundamental part of the architecture? I think a more traditional approach where one server (the back-end) handles a variable number of clients (front-ends) could have some benefits. In the following, 'client' does not necessarily mean one user or one process, but rather one view into the text, i.e. viewport, selection(s) and cursor(s).

Or do tabs implement this? Does the core share the editor state of a document among tabs of the same file, while storing client information on a per-tab basis? To me, the name 'tab' implies the UI paradigm of a text editor opening multiple tabs with (usually different) files. But not necessarily the concept of storing client state independent from the core server state. If this is the case, the documentation should probably make this more clear.

The [front-end rpc documentation](https://github.com/google/xi-editor/blob/master/doc/frontend.md) talks about tab names derived from file names as a future possibility. Does this still include the ability to open the same file in multiple tabs to allow independent cursor position, viewport and selections?

Some more situations where this could come up, and where 'tab' fails to imply this data separation:

- Making a back-end service available via d-bus, so that multiple applications can easily share the same back-end.
- Running a back-end in a network for centralized collaborative editing.
- Distributed collaborative editing where everyone runs his/her own back-end, and the server state is synchronousized via CRDTs, whereas the client-state is only kept locally.

At some point in the above examples, it also becomes important that the client-state saved on the server can only be manipulated by the correct actual client. We don't want anyone else to scroll our viewport.

- - -

To summarize: The back-end stores a global state (the text, styles etc.), as well as some connection-specific state for each client (aka tab?). This has some interesting implications for possible use-cases, and also raises the question of how to prevent access to data of other clients. The documentation does not explicitly talk about this. Tabs might deal with this, or they might not. Anyways, clarification would certainly help.
